### PR TITLE
[feat][plugin] make ATOM mla attention works for vllm

### PR DESCRIPTION
**Pull request by XiaobingSuper**
## Motivation

Following https://github.com/ROCm/ATOM/pull/126, this PR makes ATOM mla attention work for the vLLM plugin model. Note: the sparse mla is not supported now and will be implemented in the next step.

## Technical Details

The design tails can be seen in https://github.com/ROCm/ATOM/pull/126.

## Test Plan

This PR does a test for Kimi-K2-Thinking-MXFP4 mode with TP4 on mi355:

```
export SAFETENSORS_FAST_GPU=1
export VLLM_ROCM_USE_AITER=1
export VLLM_RPC_TIMEOUT=1800000

export VLLM_CACHE_ROOT=/root/.cache/vllm
export TORCHINDUCTOR_CACHE_DIR=/root/.cache/inductor
export HIP_VISIBLE_DEVICES=0,1,2,3
# quick allreduce
export AITER_QUICK_REDUCE_QUANTIZATION=INT4
export ATOM_PROFILER_MORE=1

export VLLM_TORCH_PROFILER_RECORD_SHAPES=1

model_path= Kimi-K2-Thinking-MXFP4
vllm serve $model_path \
    --host localhost \
    --port 8001 \
    --tensor-parallel-size 4 \
    --enable-expert-parallel \
    --trust-remote-code \
    --disable-log-requests \
    --gpu_memory_utilization 0.9 \
    --compilation-config '{"cudagraph_mode": "FULL_AND_PIECEWISE"}' \
    --kv-cache-dtype fp8 \
    --max-num-batched-tokens 18432 \
    --max-model-len 16384 \
    --no-enable-prefix-caching
```

## Test Result

gsmk result"
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     3|exact_match|↑  |0.9371|±  |0.0067|
|     |       |strict-match    |     3|exact_match|↑  |0.9363|±  |0.0067|
  
```
## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

File: atom/model_ops/linear.py [hunk 1]
```diff
@@ -385,8 +385,9 @@ def forward(
                 if self.quant_type.value == QuantType.per_1x128.value:
                     quant_func = functools_partial(quant_func, transpose_scale=True)
                 if self.quant_type.value != QuantType.per_1x32.value:
+                    # quant_func will call view, so we need to call contiguous to avoid view error
                     x, x_scale = quant_func(
-                        x,
+                        x.contiguous(),
```
Comment (XiaobingSuper) [inline atom/model_ops/linear.py]: This is required for the deepseek-r1 model, where x is a sliced tensor that cannot be viewed.

File: atom/model_ops/attention_mla.py [hunk 2]
```diff
@@ -577,7 +609,7 @@ def forward(
         if forward_context.context.is_dummy_run:
             # dummy run: skip real attention and return
             output_shape = list(q.shape)
-            output_shape[-1] = 7168
+            output_shape[-1] = self.num_heads * self.v_head_dim
```
Comment (XiaobingSuper) [inline atom/model_ops/attention_mla.py]: Why do we need this change? We move self.o_proj from attention to alignment with the vLLM pipeline.

File: atom/model_ops/paged_attention.py [hunk 3]
```diff
@@ -88,28 +88,85 @@ def __init__(
...
             self.layer_name = prefix
             if self.layer_name in compilation_config.static_forward_context:
                 raise ValueError("Duplicate layer: {}".format(self.layer_name))
             compilation_config.static_forward_context[self.layer_name] = self
+
+            if self.use_mla:
+                max_num_tokens = (
+                    atom_config.plugin_config.vllm_scheduler_config.max_num_batched_tokens
+                )
+                compilation_config.static_forward_context["positions"] = torch.zeros(
```
Comment (ZhangLirong-amd) [inline atom/model_ops/paged_attention.py]: hi, I remember static_forward_context will save some cls details like moe and attention, for position, our Context in ForwardContext has this variable, https://github.com/ROCm/ATOM/blob/main/atom/utils/forward_context.py#L138, can we use it?

File: atom/model_ops/attention_mla.py [hunk 4]
```diff
@@ -92,11 +97,12 @@ def dynamic_per_batched_tensor_quant(
     return x_scl_sat.to(dtype).contiguous(), scale.float().reciprocal()
 
 
+@MLAAttentionImplDecoratorForPluginMode
```
Comment (ganyi1996ppo) [inline atom/model_ops/attention_mla.py:100]: I'm not quite like the current PluginMode attention decorator. We should have better way to implement these works like isolate the attention impl from atom's original one. Maybe we can merge this first and then setup a meeting with @zejunchen-zejun to refine these part. What do you think?

Comment (zejunchen-zejun) [reply atom/model_ops/attention_mla.py:100]: we indeed need to refine the decorator

File: atom/model_ops/paged_attention.py (hunk 3 above)
Comment (XiaobingSuper) [reply atom/model_ops/paged_attention.py]: For oot path, it uses the vLLM model runner and does not use the atom model runner, meaning that the forward context will not be initialized, see https://github.com/ROCm/ATOM/blob/main/atom/model_engine/model_runner.py#L1283.

File: atom/model_ops/attention_mla.py [hunk 5]
```diff
@@ -134,7 +140,34 @@ def __init__(
...
+            self.cp_kv_cache_interleave_size: int = (
+                get_current_vllm_config().parallel_config.cp_kv_cache_interleave_size
+            )
+
+            self.is_aiter_triton_fp4_bmm_enabled = (
+                is_rocm_aiter_fp4bmm_enabled()
+                and self.kv_b_proj.weight.dtype == torch.bfloat16
+            )
+            # q_pad_num_heads in kwargs
+            self.q_pad_num_heads = kwargs.get("q_pad_num_heads", None)
```
Comment (ChuanLi1101) [inline atom/model_ops/attention_mla.py]: The __init__ signature at line 97-112 does not include **kwargs. The kwargs parameter is only available in the Attention base class constructor call chain. When is_vllm() is True, this code runs inside MLAAttention.__init__, where kwargs is not defined — this will raise NameError.
Suggested fix: Add **kwargs to MLAAttention.__init__ signature, or explicitly pass q_pad_num_heads as a named parameter.

File: atom/plugin/attention_mla.py [hunk 6]
```diff
@@ -0,0 +1,834 @@
...
+            attn_metadata.plugin_metadata.decode.max_qo_len,
+            sm_scale=self.scale,
+            work_meta_data=work_meta_data,
+            work_indptr=work_indptr,
+            work_info_set=work_info_set,
+            reduce_indptr=reduce_indptr,
+            reduce_final_map=reduce_final_map,
+            reduce_partial_map=reduce_partial_map,
+            q_scale=layer._q_scale,
+            kv_scale=layer._k_scale,
```
Comment (ChuanLi1101) [inline atom/plugin/attention_mla.py:622]: forward_decode_plugin_mode references layer._q_scale and layer._k_scale, but these are only set by process_weights_after_loading in server mode (inside the aiter_mla.py backend's prepare_decode). In plugin mode, process_weights_after_loading now takes an optional act_dtype parameter (line 72), but the scale initialization path for plugin mode is not clearly guaranteed. If these scales are None or missing, mla_decode_fwd will crash.

File: atom/model_ops/paged_attention.py [hunk 7]
```diff
@@ -88,28 +88,86 @@ def __init__(
...
+
+                    def new_forward(*args, **kwargs):
+                        out = orig_impl(*args, **kwargs)
+                        if (
+                            os.getenv("ATOM_DISABLE_VLLM_PLUGIN_ATTENTION", "0").lower()
+                            == "0"
+                        ):
+                            return out
+                        return out, None
+
```
Comment (ChuanLi1101) [inline atom/model_ops/paged_attention.py]: This reads an env var on every forward call, which is in the hot path. Also, the logic is inverted — when plugin is enabled (env=0, default), it returns out; when disabled, it returns (out, None). This seems like debug/fallback code that shouldn't be in production

File: atom/model_ops/paged_attention.py [hunk 8]
```diff
@@ -60,15 +60,15 @@ def __init__(
             **kwargs,
         )
 
+        self.use_mla = use_mla
         # for plugin mode
         if is_vllm():
-            self.use_mla = use_mla
-            self.rotary_emb = rotary_emb
+            self.rotary_emb = mla_modules.rotary_emb
```
Comment (ChuanLi1101) [inline atom/model_ops/paged_attention.py]: The original code (deleted line 303) was self.rotary_emb = rotary_emb (from the parameter). The new code uses mla_modules.rotary_emb. These should be the same object, but if use_mla=False, mla_modules may be None, and accessing mla_modules.rotary_emb will crash.

File: atom/plugin/attention_mla.py [hunk 9]
```diff
@@ -0,0 +1,834 @@
...
+            max_seqlen_q=prefill.max_query_len,
+            max_seqlen_k=prefill.max_query_len,
+            softmax_scale=self.scale,
+            causal=True,
+            return_softmax_lse=return_softmax_lse,
+        )
+
+    def _run_prefill_context_chunk_plugin_mode(self, prefill, chunk_idx, q, k, v):
+        assert prefill.chunked_context is not None
+        assert prefill.chunked_context is not None
```
Comment (ChuanLi1101) [inline atom/plugin/attention_mla.py:224]: duplicated?

File: atom/model_ops/paged_attention.py (hunk 8 above)
Comment (XiaobingSuper) [reply atom/model_ops/paged_attention.py]: Updated, for mla, rotary_emb is none, now there is a check when using mla_modules.rotary_emb.

File: atom/plugin/attention_mla.py (hunk 9 above)
Comment (XiaobingSuper) [reply atom/plugin/attention_mla.py:224]: removed.

File: atom/model_ops/paged_attention.py (hunk 7 above)
Comment (XiaobingSuper) [reply atom/model_ops/paged_attention.py]: Yes, this env check is removed.

File: atom/model_ops/attention_mla.py (hunk 5 above)
Comment (XiaobingSuper) [reply atom/model_ops/attention_mla.py]: I checked the MLAAttention.init, it has **kwargs.

File: atom/plugin/attention_mla.py (hunk 6 above)
Comment (XiaobingSuper) [reply atom/plugin/attention_mla.py:622]: For plugin mode, the scale will be initialized at https://github.com/vllm-project/vllm/blob/v0.15.1/vllm/attention/layer.py#L685, and we use it. the **self.impl.process_weights_after_loadin**g using atom mla attention methond(https://github.com/vllm-project/vllm/blob/v0.15.1/vllm/attention/layer.py#L675C13-L675C52).

Comment (XiaobingSuper) [reply atom/plugin/attention_mla.py:622]: For ATOM, path I checked the scales never get updated, always the same as vllm side scale

File: atom/plugin/attention_mla.py [hunk 10]
```diff
@@ -0,0 +1,833 @@
...
+            if self.is_aiter_triton_fp4_bmm_enabled:
+                decode_ql_nope = batched_gemm_a16wfp4(
+                    decode_q_nope,
+                    self.W_K,
+                    self.W_K_scale,
+                    transpose_bm=True,
+                    prequant=True,
+                    y_scale=layer._q_scale if fp8_attention else None,
+                )
+            # elif self.is_aiter_triton_fp8_bmm_enabled:
```
Comment (zejunchen-zejun) [inline atom/plugin/attention_mla.py:752]: what is the default value for is_aiter_triton_fp4_bmm_enabled and is_aiter_triton_fp8_bmm_enabled? Default true right?

Comment (XiaobingSuper) [reply atom/plugin/attention_mla.py:752]: yes

File: atom/model_ops/attention_mla.py (hunk 2 above)
Comment (ZhangLirong-amd) [reply atom/model_ops/attention_mla.py]: And maybe here need to change too? https://github.com/ROCm/ATOM/blob/main/atom/model_ops/base_attention.py#L195

Comment (XiaobingSuper) [reply atom/model_ops/attention_mla.py]: @ZhangLirong-amd  They don't need to change, see 
image
the unified_attention_with_output_base has self.impl.o_proj, and the output_shape[-1] is 7168.

Comment (ZhangLirong-amd) [reply atom/model_ops/attention_mla.py]: ok. Since we want to remove the hardcode 7168 there to support more mla model later

Comment (XiaobingSuper) [reply atom/model_ops/attention_mla.py]: Remove the hardcoded 7168.

File: atom/models/deepseek_v2.py [hunk 11]
```diff
@@ -1899,6 +1900,15 @@ def make_empty_intermediate_tensors(
     def get_expert_mapping(self) -> list[tuple[str, str, int, str]]:
         return self.model.get_expert_mapping()
 
+    def load_weights(self, weights: Iterable[tuple[str, torch.Tensor]]) -> set[str]:
```
Comment (valarLip) [inline atom/models/deepseek_v2.py]: refer this one? https://github.com/ROCm/ATOM/pull/262

File: atom/model_ops/attention_mla.py [hunk 12]
```diff
@@ -134,7 +140,34 @@ def __init__(
         )
         self.layer_num = layer_num
 
-    def process_weights_after_loading(self):
+        # for plugin mode(vllm)
```
Comment (valarLip) [inline atom/model_ops/attention_mla.py]: maybe do these stuff in atom/plugin/attention_mla.py

File: atom/model_ops/attention_mla.py (hunk 2 above)
Comment (valarLip) [inline atom/model_ops/attention_mla.py]: self.num_heads * self.v_head_dim looks like not eaquals to 7168 for deepseek

Comment (XiaobingSuper) [reply atom/model_ops/attention_mla.py]: This is o_proj's input,  see atom path:
image and plugin path: 
image



The reason is that vllm do a_proj outside of the attention backend.

File: atom/models/deepseek_v2.py (hunk 11 above)
Comment (XiaobingSuper) [reply atom/models/deepseek_v2.py]: Yes, updated.

File: atom/model_ops/attention_mla.py (hunk 12 above)
Comment (XiaobingSuper) [reply atom/model_ops/attention_mla.py]: moved.

File: atom/model_ops/attention_mla.py (hunk 4 above)
Comment (valarLip) [reply atom/model_ops/attention_mla.py:100]: +1... so many dirty works

File: atom/model_ops/linear.py (hunk 1 above)
Comment (valarLip) [reply atom/model_ops/linear.py]: we will do something else to avoid contiguous which will introduce mem copy here, all our quant should support non-contiguous tensor already... we hit any issue here?

File: atom/model_ops/attention_mla.py (hunk 2 above)
Comment (valarLip) [reply atom/model_ops/attention_mla.py]: plugin path is also in our repo... then why we have to move o_proj out of attn

File: atom/model_ops/linear.py (hunk 1 above)
Comment (XiaobingSuper) [reply atom/model_ops/linear.py]: Yes, I meet this issue for the chunked prefill path. see https://github.com/XiaobingSuper/ATOM/blob/xiaobing/oot_kimi/atom/plugin/attention_mla.py#L345.

File: atom/model_ops/attention_mla.py (hunk 2 above)
Comment (XiaobingSuper) [reply atom/model_ops/attention_mla.py]: This change is for the fallback path, i.e., for plugin mode, but use vllm attn backend, because we use vllm MLAAttention class(here is self.attn), the forward path doesn't has o_proj, see https://github.com/vllm-project/vllm/blob/v0.15.1/vllm/attention/layer.py#L640, this is only for attention compute.

Comment (XiaobingSuper) [reply atom/model_ops/attention_mla.py]: @valarLip The plugin mode using vllm atten backend will be like(setting ATOM_DISABLE_VLLM_PLUGIN_ATTENTION=1): https://github.com/zejunchen-zejun/ATOM/blob/zejun/plugin_for_atom_1223/recipes/vLLM-ATOM-OOT-Plugin-Backend.md#launching-server-of-vllm-with-atom-oot-plugin-platform

File: atom/model_ops/linear.py (hunk 1 above)
Comment (valarLip) [reply atom/model_ops/linear.py]: then do contiguous at that place maybe, i don't like loss any perf

Comment (XiaobingSuper) [reply atom/model_ops/linear.py]: Yes, I updated the code, and do contiguous at plugin side.

File: atom/model_ops/paged_attention.py [hunk 13]
```diff
@@ -153,7 +204,8 @@ def __init__(
             k_norm=k_norm,
             **kwargs,
         )
-
+        impl_args["head_size" if self.use_mla else "head_dim"] = head_dim
```
Comment (valarLip) [inline atom/model_ops/paged_attention.py]: this is also comes from vllm? i would like we always use head_dim

File: atom/model_ops/attention_mla.py [hunk 14]
```diff
@@ -134,7 +140,7 @@ def __init__(
         )
         self.layer_num = layer_num
 
-    def process_weights_after_loading(self):
+    def process_weights_after_loading(self, act_dtype: Optional[torch.dtype] = None):
```
Comment (valarLip) [inline atom/model_ops/attention_mla.py]: @zejunchen-zejun  we need add this arg?

File: atom/model_ops/paged_attention.py (hunk 13 above)
Comment (XiaobingSuper) [reply atom/model_ops/paged_attention.py]: Yes, vllm use head_size, see https://github.com/vllm-project/vllm/blob/v0.15.1/vllm/attention/layer.py#L579. Before this PR https://github.com/ROCm/ATOM/commit/6c402484c8f1ea3d979cc9ce8b46db8f51d990a3, it also use head_size.

File: atom/model_ops/attention_mla.py (hunk 14 above)
Comment (XiaobingSuper) [reply atom/model_ops/attention_mla.py]: For vLLM side, it's calling path like https://github.com/vllm-project/vllm/blob/1892993bc18e243e2c05841314c5e9c06a80c70d/vllm/attention/layer.py#L675, it needs such a arg.

File: atom/model_ops/paged_attention.py (hunk 13 above)
Comment (valarLip) [reply atom/model_ops/paged_attention.py]: ic……… @zejunchen-zejun please unify the args…it's begin mess now

File: atom/model_ops/attention_mla.py (hunk 2 above)
Comment (XiaobingSuper) [reply atom/model_ops/attention_mla.py]: @valarLip, currently, we only move o_proj out of attn only for plugin mode.

File: atom/model_ops/attention_mla.py (hunk 14 above)
Comment (XiaobingSuper) [reply atom/model_ops/attention_mla.py]: @valarLip, this arg is removed now, and we warp the vllm mathon in plugin mode.

Comment (zejunchen-zejun) [reply atom/model_ops/attention_mla.py]: Hi @valarLip 
The `process_weights_after_loading` is called by vLLM side and vLLM requires that the Attention needs to have `act_type` argument, but don't worry, we can use wrapper here.

**Review by zejunchen-zejun — APPROVED** [review]

File: atom/model_ops/paged_attention.py (hunk 13 above)
Comment (XiaobingSuper) [reply atom/model_ops/paged_attention.py]: @valarLip @zejunchen-zejun , now all use head_size. 1. vllm attention all use head_size, 2. atom other attention also use head_size, such as GDN.

Comment (zejunchen-zejun) [reply atom/model_ops/paged_attention.py]: Thank you @XiaobingSuper 
@valarLip It makes sense to clear the mess here, previously I mess the head_dim and head_size for MHA and MLA, for now Xiaobing cleans the code!
